File: app/ai/color_key_engine.py
import numpy as np
import cv2
from app.ai.base import BackgroundRemovalEngine
from app.utils.logger import logger

class ColorKeyEngine(BackgroundRemovalEngine):
    """
    Engine for removing backgrounds based on color similarity (Chroma Key / Lab color space).
    """

    def __init__(self):
        self._key_color = (255, 255, 255)  # RGB
        self._tolerance = 40                 # Color distance threshold
        self._feather = 5                    # Edge softness

    def load(self, model_name: str = "ColorKey", device: str = "Auto") -> bool:
        return True

    def unload(self):
        pass

    def is_loaded(self) -> bool:
        return True

    def process(self, image: np.ndarray, key_color: tuple = None, tolerance: int = 40, feather: int = 5, **kwargs) -> np.ndarray:
        if image is None or image.size == 0:
            raise ValueError("Input image is empty.")

        if key_color is None:
            key_color = self._key_color

        h, w = image.shape[:2]
        if image.shape[2] == 4:
            rgb_img = cv2.cvtColor(image, cv2.COLOR_RGBA2RGB)
        else:
            rgb_img = image.copy()

        # Convert image and key color to CIE LAB color space for perceptual accuracy
        lab_img = cv2.cvtColor(rgb_img, cv2.COLOR_RGB2LAB).astype(np.float32)
        key_rgb_np = np.uint8([[list(key_color)]])
        key_lab = cv2.cvtColor(key_rgb_np, cv2.COLOR_RGB2LAB).astype(np.float32)[0, 0]

        # Calculate Euclidean distance in Lab space
        dist = np.linalg.norm(lab_img - key_lab, axis=2)

        # Distance thresholding
        mask = np.ones((h, w), dtype=np.float32)
        bg_indices = dist <= tolerance
        mask[bg_indices] = 0.0

        # Soft edge feathering near tolerance boundary
        feather_zone = (dist > tolerance) & (dist <= tolerance + feather)
        if np.any(feather_zone) and feather > 0:
            mask[feather_zone] = (dist[feather_zone] - tolerance) / float(feather)

        mask_uint8 = (mask * 255.0).clip(0, 255).astype(np.uint8)
        return mask_uint8

File: Cpp/src/ai/ColorKeyEngine.cpp
#include "ai/ColorKeyEngine.h"
#include <opencv2/imgproc.hpp>
#include <algorithm>

namespace ImageCut {
namespace AI {

ColorKeyEngine::ColorKeyEngine() {}

bool ColorKeyEngine::load(const QString& modelName, const QString& device) {
    (void)modelName; (void)device;
    return true;
}

void ColorKeyEngine::unload() {}

bool ColorKeyEngine::isLoaded() const {
    return true;
}

cv::Mat ColorKeyEngine::process(const cv::Mat& image) {
    return processColorKey(image, m_keyColor, m_tolerance, m_feather);
}

cv::Mat ColorKeyEngine::processColorKey(const cv::Mat& image, const QColor& keyColor, int tolerance, int feather) {
    if (image.empty()) return cv::Mat();

    cv::Mat rgbImg;
    if (image.channels() == 4) {
        cv::cvtColor(image, rgbImg, cv::COLOR_RGBA2RGB);
    } else {
        rgbImg = image.clone();
    }

    cv::Mat labImg;
    cv::cvtColor(rgbImg, labImg, cv::COLOR_RGB2Lab);
    labImg.convertTo(labImg, CV_32FC3);

    cv::Mat keyRgb(1, 1, CV_8UC3, cv::Scalar(keyColor.red(), keyColor.green(), keyColor.blue()));
    cv::Mat keyLab;
    cv::cvtColor(keyRgb, keyLab, cv::COLOR_RGB2Lab);
    keyLab.convertTo(keyLab, CV_32FC3);
    cv::Vec3f keyVal = keyLab.at<cv::Vec3f>(0, 0);

    int h = image.rows;
    int w = image.cols;
    cv::Mat mask(h, w, CV_8UC1, cv::Scalar(255));

    for (int y = 0; y < h; ++y) {
        for (int x = 0; x < w; ++x) {
            cv::Vec3f val = labImg.at<cv::Vec3f>(y, x);
            float dist = std::sqrt(
                (val[0] - keyVal[0]) * (val[0] - keyVal[0]) +
                (val[1] - keyVal[1]) * (val[1] - keyVal[1]) +
                (val[2] - keyVal[2]) * (val[2] - keyVal[2])
            );

            if (dist <= tolerance) {
                mask.at<uint8_t>(y, x) = 0;
            } else if (dist <= tolerance + feather && feather > 0) {
                float alpha = (dist - tolerance) / static_cast<float>(feather);
                mask.at<uint8_t>(y, x) = static_cast<uint8_t>(alpha * 255.0f);
            }
        }
    }

    return mask;
}

} // namespace AI
} // namespace ImageCut
